Add parameter to set user uuid in jenkins::credentials define

madAndroid opened the pull request:
This addresses #226 - It's largely influenced by the Opscode jenkins module, which has this functionality - https://github.com/opscode-cookbooks/jenkins/blob/master/libraries/credentials.rb

madAndroid commented:
@rtyler - Any progress on testing this? :) Please let me know if more tests are required

xiii commented:
@rtyler @madAndroid that would be very useful! hope it gets merged soon!

jhoblitt commented:
The last CI run failed for this PR and it needs to be rebased on the current master due to changes introduced by #285.

madAndroid commented:
Cool, no worries - I'll rebase and sort out any conflicts

madAndroid commented:
Rebased, and tests fixed :) ( https://travis-ci.org/jenkinsci/puppet-jenkins/builds/63873665 )

jhoblitt commented on files/puppet_helper.groovy:
```diff
@@ -170,10 +170,13 @@ class Actions {
     def credentials
+    if (id == "") {
+      id = null
+    }
```
From looking at the existing code, I thought that the cli argument to method call conversion, `actions."$action"(*args[1..-1])`, must be requiring string params to be defaulted to an empty string.  However, I just tested it and there doesn't appear to be an issue with defaulting a string param to `null`.  

The `id` param could be defaulted to `null` and then this check wouldn't be needed.

jhoblitt replied:
Actually, that might not make any sense.  I'm not sure how you would pass null from the cli.

madAndroid replied:
> I'm not sure how you would pass null from the cli.

That's pretty much the issue I came up against, which is why I ended up taking this route.

jhoblitt commented:
:+1: This looks good to me!

CI didn't run; travis has been acting strange the last few days.  I've had to force push some of my own branches in order to get it trigger.

madAndroid commented:
Odd, I did see the run on travis after I pushed the latest changes up .. results are here, AFAICT ->  https://travis-ci.org/jenkinsci/puppet-jenkins/builds/63873665

madAndroid commented:
An example of the use case where you may want to set the UUID during credential creation: [JJB's SCM plugin](http://docs.openstack.org/infra/jenkins-job-builder/scm.html#scm.git) allows you to set `credential-id` field, which allows you to have a predefined UUID for easy rebuild of envs

madAndroid commented:
Is there anything more I can do to have this request merged? ( @jhoblitt / @rtyler )

jhoblitt commented:
This looks pretty good to me.  Mentioning the new param in the README might not hurt. (eg., https://github.com/jenkinsci/puppet-jenkins#credentials)

One thing of note is that this does change the order of params to `create_or_update_credential` -- I'm not sure if that's ever been considered a public API, @rtyler ?

madAndroid commented:
@rtyler - anything more I need to do to have this merged? I've added tests, would you like me to expand upon those?

madAndroid commented:
Hi, is there anything further than can be done to have this merged? Do you need me to justify the reason for the addition, and perhaps explain and describe our use case? 

@jhoblitt - in response to your query: the ordering of args in `create_or_update_credential` is taken directly from the credentials-plugin source: https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java#L56-L64
ie:
```
    /**
     * Constructor.
     *
     * @param scope       the credentials scope
     * @param id          the ID or {@code null} to generate a new one.
     * @param description the description.
     * @param username    the username.
     * @param password    the password.
     */
```
I've rebased against the latest HEAD of master, and fixed tests that changed as a result of another merge.

rtyler commented:
@madAndroid nope, thanks for the bump and the good contribution :)

madAndroid commented:
Awesome, thanks! :-)
On 9 Aug 2015 5:58 pm, "R. Tyler Croy" <notifications@github.com> wrote:

> @madAndroid <https://github.com/madAndroid> nope, thanks for the bump and
> the good contribution :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/puppet-jenkins/pull/227#issuecomment-129204220>
> .
>
